---
title: "pool_analysis"
author: "Rose-Marie Fraboulet"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **Stage M1 - Analyse de l’hétérogénéité tumorale du cancer du pancréas par des approches single-cell**

## **Loading packages**
```{r , include=FALSE}
#Data
library(data.table)
library(dplyr) 
library(writexl)

#Seurat
library(Seurat)
library(patchwork) #plots
library(ggplot2)
library(umap)

#scCatch
#install.packages(pkgs = 'devtools')
#devtools::install_github('ZJUFanLab/scCATCH')
library(scCATCH)

#inferCNV
library(rjags)

#if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#BiocManager::install("infercnv")
library(infercnv)

library(dplyr)
library(matrixStats)

#ComplexHeatMap
library(devtools)
library(usethis)

#install_github("jokergoo/ComplexHeatmap")
library(ComplexHeatmap)
library(circlize)
```

## **Loading dataset**
```{r , include=FALSE}
load("/home/rose/Documents/projet_stage/data/data/pool_object.RData")
```

## **Standard pre-processing workflow - Data filtering**
### Calculate the percentage of mitochondrial genes 
```{r, echo=FALSE}
pool[["percent.mt"]] <- PercentageFeatureSet(object = pool, pattern = "^MT-")
#number of mitochondrial genes for one cell / total number of genes in that cell * 100

head(x = pool@meta.data, 10)
#The total number of molecules detected within a cell : nCounts
#The number of genes detected in each cell : nFeatures (nGenes)
#The percentage of reads that map to the mitochondrial genome : percent.mt 
```

### Graphical visualization 
#### Violin plot
```{r, echo=FALSE}
VlnPlot(pool, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

#### Plot
```{r include=FALSE}
plot1 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot1

plot2 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2
```

### Low quality cells
<200 genes/cell and >10% mitochondrial genes were excluded, already done by Peng et al. 
```{r, echo=FALSE}
pool <- subset(pool, subset = nFeature_RNA > 200 & percent.mt < 10)
pool
```

## **Normalizing data**
## Normalizing the data
```{r, echo=FALSE}
#Normalizes the gene expression for each cell by the total expression, multiplies this by a scale factor (10 000) and log-transforms the result.
pool <- NormalizeData(pool, normalization.method = "LogNormalize", scale.factor=10000)
pool
```

## Identification of highly variable features
```{r, echo=FALSE}
#Calculates highly variable genes - Highly expressed in some cells, and lowly expressed in others
pool <- FindVariableFeatures(object=pool, selection.method = "vst", nfeatures = 2000)
head(x = HVFInfo(object = pool))

#The 10 most higly variable genes
top10_variablegenes <- head(VariableFeatures(pool), 10)
top10_variablegenes

#plot3 <- VariableFeaturePlot(pool)
#plot3

#plot4 <- LabelPoints(plot = plot3, points = top10_variablegenes, repel = TRUE, xnudge=0, ynudge=0)
#plot4
```

## **Scaling data**
```{r, include=FALSE}
# Pre-processing step before PCA
pool <- ScaleData(pool)
```

## **Perform linear dimensional reduction**
### Dataset
```{r, include=FALSE}
pool <- RunPCA(pool, features = VariableFeatures(object = pool))
#print(pool[["pca"]], dims = 1:5, nfeatures = 5)
```

### Visualization
#### VizDimReduction
```{r, include=FALSE}
VizDimLoadings(pool, dims = 1:2, reduction = "pca")
```

#### DimPlot
```{r, include=FALSE}
DimPlot(pool, reduction = "pca", label=TRUE, repel=TRUE)
```

#### DimHeatMap
```{r, echo=FALSE}
DimHeatmap(pool, dims = 1:3, cells = 500, balanced = TRUE)
#black = 0, purple = low expression, yellow = high expression
```

## **Determine the ‘dimensionality’ of the dataset**
```{r, echo=FALSE}
ElbowPlot(object=pool, ndims = 50, reduction = "pca")
```

## **Cluster the cell**
### Shared Nearest Neighbor (SNN) Graph
```{r, echo=FALSE}
#Determine the k-nearest neighbors of each cell
pool <- FindNeighbors(pool, reduction='pca', dims = 1:19)
pool <- FindClusters(pool, resolution = 1) 
```

#### Visualization 
```{r, echo=FALSE}
head(Idents(pool), 10)

#Which cluster has been determined for each cell
head(pool@meta.data, 10)
```

## **Run non-linear dimensional reduction**
### Uniform Manifold Approximation and Projection (UMAP) 
```{r, echo=FALSE}
pool <- RunUMAP(pool, dims = 1:19)
```

## **Finding differentially expressed features (cluster biomarkers)**
```{r, echo=FALSE}
pool_markers <- FindAllMarkers(pool, only.pos = TRUE)
pool_markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
head(pool_markers)
```

### Visualization 
#### Feature plot
```{r}
FeaturePlot(pool,  label=TRUE, repel=TRUE)
```

#### Top 100 differentially expressed genes for each cluster
```{r, results="hide"}
top100_markers <- pool_markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_logFC)
head(top100_markers)
```

## **Addmodulescore**
```{r}
load("~/Documents/projet_stage/data/signatures/geneSignatures.RData")

pool <- AddModuleScore(pool, ImmunePop, name="module_score_ImmunePop_Sig")
pool <- AddModuleScore(pool, otherSig, name="module_score_other_Sig")
head(pool@meta.data)

#plot4 <- FeaturePlot(object=pool, features="module_score_ImmunePop_Sig1", label=TRUE, repel=TRUE)
#plot4
#plot3 <- VlnPlot(pool, features="module_score_ImmunePop_Sig1", pt.size = FALSE) + theme(legend.position='none')
#plot4+plot3

#Average of add_module_score
metadata <- pool@meta.data
matrix <- aggregate(list(metadata$module_score_other_Sig1, metadata$module_score_other_Sig2, metadata$module_score_other_Sig3, metadata$module_score_other_Sig4, metadata$module_score_ImmunePop_Sig1, metadata$module_score_ImmunePop_Sig2, metadata$module_score_ImmunePop_Sig3, metadata$module_score_ImmunePop_Sig4, metadata$module_score_ImmunePop_Sig5, metadata$module_score_ImmunePop_Sig6, metadata$module_score_ImmunePop_Sig7, metadata$module_score_ImmunePop_Sig8, metadata$module_score_ImmunePop_Sig9, metadata$module_score_ImmunePop_Sig10), by=list(metadata$seurat_clusters), FUN=median)
```

## **Assigning cell type identity to clusters**
```{r, echo=FALSE}
Idents(pool) <- "seurat_clusters"
experiment.merged <- RenameIdents(object = pool, '30'= '0', '31'= '0', '26'= '0', '6' = '1', '8' = '1', '9'= '2', '20'= '2', '25'= '2', '24'= '2', '25'= '2', '18'= '2', '14'= '2', '13'= '3', '29'= '3', '27'= '4', '34'= '4', '12'= '4', '5'= '4', '33'= '4', '32'= '4', '17'='22', '11'='10', '15'='10', '21'='10', '16'='19', '28'='19')

DimPlot(object = pool, pt.size=0.5, label = T, reduction = "umap")
#DimPlot(object = experiment.merged, pt.size=0.5, label = T, reduction = "tsne")

new.cluster.ids <- c("Stellate cell", "Ductal cell 1", "Ductal cell 2", "Macrophage", "Endothelial cell", "Acinar cell", "Fibroblast", "T cell", "B cell", "Endocrine cell")
names(new.cluster.ids) <- levels(experiment.merged)
pool <- RenameIdents(experiment.merged, new.cluster.ids)

DimPlot(pool, reduction = "umap", pt.size = 0.5, repel=TRUE, label= TRUE) + NoLegend()
```

## **scCATCH**
```{r, echo=FALSE}
clu_markers_sccatch <- findmarkergenes(pool, species = 'Human', cluster = 'All', match_CellMatch = TRUE, cancer = NULL, tissue = 'Pancreas', cell_min_pct = 0.25, logfc = 0.25, pvalue = 0.05)
#large datasets > 10,000 cells or > 15 clusters

clu_ann_sccatch <- scCATCH(clu_markers_sccatch$clu_markers, species = 'Human', cancer = NULL, tissue = 'Pancreas')
```

## **inferCNV**
### Running
```{r, echo=FALSE}
infercnv_obj = CreateInfercnvObject(raw_counts_matrix="singleCell.counts.matrix",
annotations_file="metadata_pool_infercnvtab",
delim="\t",
gene_order_file="gen_order.txt",
ref_group_names=NULL,
min_max_counts_per_cell=c(1e4,1e5))

infercnv_obj = infercnv::run(infercnv_obj,
  cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
  out_dir="output_dirHMM",  # dir is auto-created for storing outputs
  cluster_by_groups=T,   # cluster
  denoise=T,
  HMM=T
)
```

### Data treatment
```{r, echo=FALSE}
inferCNVNormExp <- as.data.frame(inferCNVNormExp)

#normal samples
infer_CNV_N <- select(inferCNVNormExp, starts_with("N"))
colnames(infer_CNV_N)

infer_CNV_N <- abs(infer_CNV_N) #absolute value
infer_CNV_N <- as.matrix(infer_CNV_N)
infer_CNV_N_colMeans <- colMeans(infer_CNV_N)

quantile_N <- quantile(infer_CNV_N_colMeans, probs=0.99)
quantile_N 
#0.0212167 

#tumoral samples
infer_CNV_T <- select(inferCNVNormExp, starts_with("T"))
colnames(infer_CNV_T)

infer_CNV_T <- abs(infer_CNV_T)
infer_CNV_T <- as.matrix(infer_CNV_T)
infer_CNV_T_colMeans <- colMeans(infer_CNV_T)
#infer_CNV_T_colMeans <- as.data.frame(infer_CNV_T_colMeans)

infer_CNV_T_colMeans$CNA <- infer_CNV_T_colMeans$infer_CNV_T

infer_CNV_T_colMeans$CNA[infer_CNV_T_colMeans$CNA > quantile_N] <- "CNA+"
sum(infer_CNV_T_colMeans$CNA == "CNA+")
#4161 : CNA +
infer_CNV_T_colMeans$CNA[infer_CNV_T_colMeans$CNA <= quantile_N] <- "CNA-"
infer_CNV_T_colMeans[3566, 2] <- "CNA-"
sum(infer_CNV_T_colMeans$CNA == "CNA-")
#9041 : CNA -

infer_CNV_T_colMeans <- cbind("cell" = rownames(infer_CNV_T_colMeans), infer_CNV_T_colMeans)
rownames(infer_CNV_T_colMeans) <- 1:nrow(infer_CNV_T_colMeans)
```

