---
title: "Projet_Stage"
author: "Rose-Marie Fraboulet"
output: html_document
---

# **Stage M1 - Analyse de l’hétérogénéité tumorale du cancer du pancréas par des approches single-cell**

#```{r setup, include=FALSE} 
#knitr::opts_chunk$set(echo=TRUE)
#```

## **Loading packages**
```{r loading_packages, include=FALSE}
packageVersion("cellasign")
version[['version.string']]

library(data.table) #large data
library(dplyr) #data manipulation

library(Seurat)
library(patchwork) #plots
library(ggplot2)
library(umap)
library(tsne)

library("writexl")

#scCatch
install.packages(pkgs = 'devtools')
devtools::install_github('ZJUFanLab/scCATCH')
library(scCATCH)

#cellasign 
##install.packages("tensorflow")
##install.packages("reticulate")
library(reticulate)
##py_install("tensorflow", pip=T)
library(tensorflow)
##install_tensorflow(extra_packages = "tensorflow-probability")
#devtools::install_github("Irrationone/cellassign")
library(SingleCellExperiment)
library(cellassign)

#singlecellexperiment
library(scater)

#inferCNV
#install.packages("rjags")
library(rjags)

#inferCNV package from Bioconductor
# if (!requireNamespace("BiocManager", quietly = TRUE))
#      install.packages("BiocManager")
# BiocManager::install("infercnv")
library(infercnv)
```
```

## **Loading dataset**
### Pool
```{r, echo=FALSE}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/code/projet.RData")
#load("/home/rose/Documents/projet_stage/data/projet.RData")
load("/home/rose/Documents//projet_stage/data/data/pool_object.RData")
```

### Pool + metadata (labels) / Pool2
```{r, echo=FALSE}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre #2/Stage M1/Projet/data/peng_data")

#metadata <- fread("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/peng_data/all_celltype.txt", data.table=FALSE)

#Rows are cell names and the columns are additional metadata fields.
##rownames(metadata) <- metadata$cell.name
##head(metadata)
##metadata$cell.name <- NULL

#pool2 <- CreateSeuratObject(counts = counts, meta.data = metadata)
```

### Pool.init 
```{r}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/init.RData")
#load("/home/rose/Documents//projet_stage/data/init.RData")
```

### All objects
```{r}
#load("/home/rose/Documents//projet_stage/data/data_clusters.RData")
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/code/data_clusters.RData")
```

## **Standard pre-processing workflow - Data filtering**
### Calculate the percentage of mitochondrial genes 
```{r}
pool[["percent.mt"]] <- PercentageFeatureSet(object = pool, pattern = "^MT-")
#[[""]] : add a column "percent.mt" in the pool object 
#number of mitochondrial genes for one cell / total number of genes in that cell * 100
```

```{r}
head(x = pool@meta.data, 10)
#The total number of molecules detected within a cell : nCount (nUMIs)
#The number of genes detected in each cell : nFeature (nGenes)
#The percentage of reads that map to the mitochondrial genome : percent.mt 
```

### Graphical visualization 
#### Violin plot
```{r}
VlnPlot(pool, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

#### Plot
```{r}
plot1 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot1

plot2 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2
```

#### Low quality cells
<200 genes/cell and >10% mitochondrial genes were excluded 
```{r}
pool <- subset(pool, subset = nFeature_RNA > 200 & percent.mt < 10)
pool
```

## **Normalizing data - logNormalize**

## Normalizing the data
```{r}
#Normalizes the gene expression for each cell by the total expression, multiplies this by a scale factor (10 000) and log-transforms the result.
pool <- NormalizeData(pool, normalization.method = "LogNormalize", scale.factor=10000)
```

## Identification of highly variable features
```{r}
#Calculates highly variable genes - Highly expressed in some cells, and lowly expressed in others
pool <- FindVariableFeatures(object=pool, selection.method = "vst", nfeatures = 2000)
head(x = HVFInfo(object = pool))

#The 10 most higly variable genes
top10.variablegenes <- head(VariableFeatures(pool), 10)
top10.variablegenes

#plot3 <- VariableFeaturePlot(pool)
#plot3

plot4 <- LabelPoints(plot = plot3, points = top10_variablegenes, repel = TRUE, xnudge=0, ynudge=0)
plot4
```

## **Scaling data**
```{r}
# Pre-processing step before PCA
pool <- ScaleData(pool)
```

## **Perform linear dimensional reduction**
### Dataset
```{r}
pool <- RunPCA(pool, features = VariableFeatures(object = pool))
#print(pool[["pca"]], dims = 1:5, nfeatures = 5)
```

### Visualization
#### VizDimReduction
```{r}
VizDimLoadings(pool, dims = 1:2, reduction = "pca")
```

#### DimPlot
```{r}
DimPlot(pool, reduction = "pca", label=TRUE, repel=TRUE)
```

#### DimHeatMap
```{r}
DimHeatmap(pool, dims = 1:3, cells = 500, balanced = TRUE)
#black = 0, purple = low, yellow = high expression
```

## **Determine the ‘dimensionality’ of the dataset**
```{r}
ElbowPlot(object=pool, ndims = 50, reduction = "pca")
```

## **Cluster the cell**
### Shared Nearest Neighbor (SNN) Graph
```{r}
#Determine the k-nearest neighbors of each cell
#PCs 1 to 10 were used for graph-based clustering

pool <- FindNeighbors(pool, reduction='pca', dims = 1:10)

pool <- FindClusters(pool, resolution = 1) 
#Res = 0.8 for PDAC samples 
#Res = 1 for pooled cells from PDACs and controls 
```

#### Visualization 
```{r}
head(Idents(pool), 10)

#Which cluster has been determined for each cell
head(pool@meta.data, 10)

#Delete a column :  pool[['RNA_snn_res.1.5']] <- NULL
```

## **Run non-linear dimensional reduction**

### Uniform Manifold Approximation and Projection (UMAP) 
```{r, include=FALSE}
pool <- RunUMAP(pool, dims = 1:10)
```

### T-distributed stochastic neighbor embedding (tSNE)
```{r, results="hide"}
# pool <- RunTSNE(pool, dims.use = 1:10)
# plot8 <- DimPlot(object = pool, reduction = "tsne", label=TRUE, repel=TRUE, group.by="orig.ident")
# plot9 <- DimPlot(object = pool, reduction = "tsne", label=TRUE, repel=TRUE, group.by="cluster") + NoLegend()
# plot8 + plot9
```

## **Finding differentially expressed features (cluster biomarkers)**

### **Using markers cited in the article**
#### Find markers for cluster 1
```{r}
cluster1_markers <- FindMarkers(pool, ident.1 = 1)
head(cluster1_markers, n = 10)
#pct.1 = quel pourcentage des cellules dans la condition 1 montrent une certaine expression pour ce gène
#pct.2 = quel pourcentage des cellules dans la condition 2 montrent une certaine expression pour ce gène
#avg_logFC = combien l'expression de ce gène est plus ou moins élevée dans le cluster choisi entre les conditions
```

#### Find markers for every cluster compared to all remaining cells
```{r}
pool_markers <- FindAllMarkers(pool, only.pos = TRUE)
pool_markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
head(pool_markers)
```

#### Visualization 
```{r}
#For a specific marker
feature <- c("KRT19")
```

##### Violin plot
```{r}
VlnPlot(pool, features = feature, pt.size= FALSE, group.by="seurat_clusters")
```

##### Feature plot
```{r}
FeaturePlot(pool,  label=TRUE, repel=TRUE)
#FeaturePlot(pool, features = feature, reduction = "tsne", label=TRUE, repel=TRUE)
```

##### DoHeatMap
```{r}
DoHeatmap(subset(pool, downsample = 10), features=feature, size=3)
```

#### Top 100 differentially expressed genes for each cluster
```{r, results="hide"}
top100.markers <- pool_markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_logFC)
head(top100.markers)
#DoHeatmap(pool, features = top100_markers$gene, size=3)
```

```{r}
# .xlsx file of the top 100 markres for each cluster

#write_xlsx(top100.markers, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/Data/top100_markers.xlsx")
```

#### **Assigning cell type identity to clusters**
```{r}
Idents(pool) <- "seurat_clusters"
experiment.merged <- RenameIdents(object = pool, '30'= '0', '31'= '0', '26'= '0', '6' = '1', '8' = '1', '9'= '2', '20'= '2', '25'= '2', '24'= '2', '25'= '2', '18'= '2', '14'= '2', '13'= '3', '29'= '3', '27'= '4', '34'= '4', '12'= '4', '5'= '4', '33'= '4', '32'= '4', '17'='22', '11'='10', '15'='10', '21'='10', '16'='19', '28'='19')

DimPlot(object = pool, pt.size=0.5, label = T, reduction = "umap")
#DimPlot(object = experiment.merged, pt.size=0.5, label = T, reduction = "tsne")

new.cluster.ids <- c("Stellate cell", "Ductal cell 1", "Ductal cell 2", "Macrophage", "Endothelial cell", "Acinar cell", "Fibroblast", "T cell", "B cell", "Endocrine cell")
names(new.cluster.ids) <- levels(experiment.merged)
pool <- RenameIdents(experiment.merged, new.cluster.ids)

DimPlot(pool, reduction = "umap", pt.size = 0.5, repel=TRUE, label= TRUE) + NoLegend()
#DimPlot(pool, reduction = "tsne", pt.size = 0.5, repel=TRUE, label= TRUE) + NoLegend()

#add metadata_peng_all to pool object 
#metadata <- fread("~/Documents/projet_stage/data/all_celltype.txt")
#metadata <- as.data.frame(metadata)
#row.names(metadata) <- metadata$cell.name
#metadata$cell.name <- NULL

#pool.init <- AddMetaData(pool.init, metadata = metadata, col.name="peng_all_annotations")
#pool <- pool.init
```

### **Clustering tools - pool.init**
```{r}
#Using pool.init = pool data without the new clusters id found by known markers
```

#### **BuildClusterTree**
```{r}
#cluster_tree = BuildClusterTree(pool.init)
#Tool(object = cluster_tree, slot = 'BuildClusterTree')
#PlotClusterTree(object = cluster_tree)

#library(clustree)
#head((pool.init@meta.data))
#clustree(pool.init)
```

#### **scCATCH**
```{r}
#penser à faire normalisation, scale, runPCA, findneighbors et findclusters AV de run ces commandes : pre-processing 
clu_markers_sccatch <- findmarkergenes(pool, species = 'Human', cluster = 'All', match_CellMatch = TRUE, cancer = NULL, tissue = 'Pancreas', cell_min_pct = 0.25, logfc = 0.25, pvalue = 0.05)
#large datasets > 10,000 cells or > 15 clusters

clu_ann_sccatch <- scCATCH(clu_markers_sccatch$clu_markers, species = 'Human', cancer = NULL, tissue = 'Pancreas')

write_xlsx(clu_ann_sccatch, "/home/rose/Documents/projet_stage/data/scCATCH/clu_ann_sccatch.xlsx")
```

#### **cellasign**
###### Data import
```{r}
#load("/home/rose/Documents/projet_stage/data/sizefactors.RData")

## Create SingleCellExperiment object by counts matrix
#counts <- data.matrix(counts, rownames.force = NA) #matrix
#metadata2 <- data.matrix(metadata, rownames.force = NA) #cluter (0 to 34)
#sce <- SingleCellExperiment(list(counts=counts), colData=DataFrame(metadata2))
#print(sce)

#Create SingleCellExperiment objet using pool seurat object
pool.sce <- as.SingleCellExperiment(pool)
pool.sce <- pool.sce[,1:100]
pool.sce@colData
head(pool.sce$peng_all_annotations)
dim(pool.sce)

#Find all markers for pool
Idents(pool) <- "peng_all_annotations"
pool_markers_test <- FindAllMarkers(pool, only.pos = TRUE) 
#write_xlsx(pool_markers_test, "~/Documents/projet_stage/data/cellasign/pool_markers_test.xlsx")
load("/home/rose/Documents/projet_stage/data/cellasign/pool_markers_test.RData")

boxplot(avg_logFC ~ cluster, pool_markers_test)
pool_markers_test_avg <- subset(pool_markers_test, avg_logFC >=1.5)
head(pool_markers)
pool_markers_test_avg <- pool_markers_test_avg[6:7]

#Creation of the list, where each entry is named by a cell type and contains a character vector of gene names belonging to that cell type
split_pool_markers <- split(pool_markers_test_avg[2], pool_markers_test_avg[[1]])
str(head(split_pool_markers))

unlist_pool_markers <- unlist(split_pool_markers)
unlist_pool_markers <- unlist(split_pool_markers, use.names=FALSE)
head(unlist_pool_markers)

list_pool_markers <- lapply(split_pool_markers, function(x) unlist(x, use.names=FALSE))
str(head(list_pool_markers))  

#Creation of the marker gene matrix
mat_pool_markers <- marker_list_to_mat(list_pool_markers, include_other = FALSE)
mat_pool_markers[1:10,]

frame_mat_pool_markers <- as.data.frame(mat_pool_markers)
write_xlsx(frame_mat_pool_markers, "~/Documents/projet_stage/data/cellasign/mat_pool_markers.xlsx")
```

###### cellasign tool
```{r}
#Size factor for each cell 
library(scran)
pool.sce <- computeSumFactors(pool.sce)
head(sizeFactors(pool.sce))
s <- sizeFactors(pool.sce)
s

pool.sce
load("~/Documents/projet_stage/data/cellasign/cellasign.RData")

fit <- cellassign(exprs_obj=pool.sce[rownames(mat_pool_markers),], marker_gene_info=mat_pool_markers, s=s, learning_rate = 1e-2, shrinkage=TRUE, verbose=FALSE) 

fit_1_50 <- fit
print(fit_1_50)
print(celltypes(fit_1_50))
heatmap_1_50 <- pheatmap::pheatmap(cellprobs(fit_1_50), legend=TRUE, main="fit_1_50")
heatmap_25_75 <- pheatmap::pheatmap(cellprobs(fit_25_75), legend=TRUE, main="fit_25_75")
heatmap_50_100 <- pheatmap::pheatmap(cellprobs(fit_50_100), legend=TRUE, main="fit_50_100")
```

##example cellasign
```{r}
data(example_sce)
data(example_marker_mat)
s <- sizeFactors(example_sce)

fit <- cellassign(exprs_obj = example_sce[rownames(example_marker_mat),], 
                  marker_gene_info = example_marker_mat, 
                  s = s, 
                  learning_rate = 1e-2, 
                  shrinkage = TRUE,
                  verbose = FALSE)

pheatmap::pheatmap(cellprobs(fit))                
```


### **Using cell labels - pool2**
```{r}
plot5 <- DimPlot(pool2, reduction = "umap", label=TRUE, repel=TRUE, group.by="orig.ident")
plot5

Idents(pool) <- "orig.ident"
experiment.merged2 <- RenameIdents(object = pool, 'N11'='N1', 'N10'='N1', 'N9'='N1', 'N8'='N1', 'N7'='N1', 'N6'='N1', 'N5'='N1', 'N4'='N1', 'N3'='N1', 'N2'='N1', 'T2'='T1', 'T3'='T1','T4'='T1','T5'='T1','T6'='T1','T7'='T1','T8'='T1','T9'='T1','T10'='T1','T11'='T1','T12'='T1','T13'='T1','T14'='T1','T15'='T1','T16'='T1','T17'='T1','T18'='T1','T19'='T1','T20'='T1','T21'='T1','T22'='T1','T23'='T1','T24'='T1')
#plot6 <-DimPlot(object = experiment.merged2, pt.size=0.5, label = T, reduction = "umap")
new.orig_ident.ids <- c("Control samples", "Cancer samples")
names(new.orig_ident.ids) <- levels(experiment.merged2)
pool<- RenameIdents(experiment.merged2, new.orig_ident.ids)

plot6 <- DimPlot(pool, reduction = "umap", pt.size = 0.5, cols=c("seagreen2", "firebrick3"))
plot6 

plot7 <- DimPlot(pool2, reduction = "umap", label=TRUE, repel=TRUE, group.by="cluster") + NoLegend()
plot7
```

### **Calculating the average gene expression within a cluster**
```{r}
Idents(pool2) <- "cluster"
#Idents(pool2) <- "orig.ident"

#How many cells are in each cluster
table(Idents(pool2))

#How many cells are in each sample
prop.table(pool2$orig.ident)

pool2$orig.ident <- as.numeric(as.character(pool2$orig.ident))
#How many cells per sample
table(Idents(pool2), pool2$orig.ident)

cluster.averages <- AverageExpression(pool2)
head(cluster.averages[["RNA"]][, 1:5])

#Add cluster.averages to the pool2 object
orig.levels <- levels(pool2)
Idents(pool2) <- gsub(pattern = " ", replacement = "_", x = Idents(pool2))
orig.levels <- gsub(pattern = " ", replacement = "_", x = orig.levels)
levels(pool2) <- orig.levels
cluster.averages <- AverageExpression(pool2, return.seurat = TRUE)
cluster.averages

CellScatter(cluster.averages, cell1 = "Ductal_cell_type_1", cell2 = "Ductal_cell_type_2")

DoHeatmap(cluster.averages, features = unlist(TopFeatures(pool2[["pca"]], balanced = TRUE)), size = 3, 
    draw.lines = FALSE)
```


```{r}
##Contengency table
#orig.ident <- pool2$orig.ident
#clusters <- pool.init$seurat_clusters

#contengency.table <- data.frame(orig.ident, clusters)

#table <- data.frame(metadata, orig.ident, clusters)

#write_xlsx(contengency.table, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/supplementary_information/contengency_table.xlsx")

#write_xlsx(table, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/supplementary_information/table.xlsx")
```

### **inferCNV**
##### Loading files
```{r}
#counts matrix
load("~/Documents/projet_stage/data/data/Peng_Data.RData")
#write.table(counts, file="~/Documents/projet_stage/data/inferCNV/counts.txt") 

#annotations file 

#with annotation for cell type
#metadata <- fread("~/Documents/projet_stage/data/all_celltype.txt")
#metadata <- as.data.frame(metadata)
names(metadata) <- NULL
head(metadata)
#write.table(metadata, file="~/Documents/projet_stage/data/inferCNV/metadata.txt", quote=F, sep="\t", row.names=F)

#with cluster numbers
#metadata_pool_infercnv <- metadata_pool
#metadata_pool_infercnv <- metadata_pool_infercnv[,-(2:7)]
#names(metadata_pool_infercnv)[names(metadata_pool_infercnv) == "seurat_clusters"] <- "cluster"
#write.table(metadata_pool_infercnv, file="~/Documents/projet_stage/data/metadata_pool_infercnv", sep="\t", row.names=F, col.names=T, quote=F)      
#row.names(metadata_pool_infercnv) <- metadata_pool_infercnv$cell
#names(metadata_pool_infercnv) <- NULL
#metadata_pool_infercnv <- metadata_pool_infercnv[,-1]
metadata_pool_infercnv <- fread("~/Documents/projet_stage/data/metadata_pool_infercnv")

#gene order file 
gene_list <- data.frame(row.names(counts))
gen_code <- read.delim("~/Documents/projet_stage/data/gencode_v19_gene_pos.txt",header=F)
names(gen_code) <- c("gene", "chr", "gene span1", "gene span2")
names(gene_list) <- c("gene")

gen_order <- merge(gene_list, gen_code, by="gene")
head(gen_order)
#loss of 170 genes 

#no header
names(gen_order) <- NULL
head(gen_order)

#write.table(gen_order, file="~/Documents/projet_stage/data/inferCNV/gen_order.txt", quote=FALSE, sep="\t", row.names=F)

load("~/Documents/projet_stage/data/inferCNV/inferCNV.RData")
```

##### inferCNV 
```{r}
#Create the inferCNV object 
infercnv_objt <- CreateInfercnvObject(counts, annotations_file="~/Documents/projet_stage/data/metadata_pool_infercnv", gene_order_file ="~/Documents/projet_stage/data/inferCNV/gen_order.txt", ref_group_names=NULL)

#Run the standard infercnv procedure
infercnv_objt = infercnv::run(infercnv_objt,
                             cutoff=0, 
                             out_dir=tempfile(), 
                             cluster_by_groups=FALSE, 
                             denoise=TRUE,
                             HMM=TRUE,
                             noise_filter = 0.2)
#session aborted at step 2 : filtering genes 

#cf résultats 
load("~/Documents/projet_stage/data/inferCNV/results_remy/InferCNVNormalizedCounts.RData")
#write.table(inferCNVNormExp, file="~/Documents/projet_stage/data/inferCNV/results_remy/InferCNVNormalizedCounts.xlsx", sep="\t")
```


```{r}
library(dplyr)
library(matrixStats)
#Data treatment of normal and tumor samples from inferCNV matrix 

inferCNVNormExp <- as.data.frame(inferCNVNormExp)

#normal samples
infer_CNV_N <- select(inferCNVNormExp, starts_with("N"))
colnames(infer_CNV_N)

infer_CNV_N <- abs(infer_CNV_N) #absolute value
infer_CNV_N <- as.matrix(infer_CNV_N)
infer_CNV_N_colMeans <- colMeans(infer_CNV_N)
#infer_CNV_N_colMeans <- as.data.frame(infer_CNV_N_colMeans)
#infer_CNV_N_colMeans[1:3]

quantile_N <- quantile(infer_CNV_N_colMeans, probs=0.99)
quantile_N 
#0.0212167 


#tumoral samples
infer_CNV_T <- select(inferCNVNormExp, starts_with("T"))
colnames(infer_CNV_T)

infer_CNV_T <- abs(infer_CNV_T)
infer_CNV_T <- as.matrix(infer_CNV_T)
infer_CNV_T_colMeans <- colMeans(infer_CNV_T)
#infer_CNV_T_colMeans <- as.data.frame(infer_CNV_T_colMeans)

infer_CNV_T_colMeans$CNA <- infer_CNV_T_colMeans$infer_CNV_T

infer_CNV_T_colMeans$CNA[infer_CNV_T_colMeans$CNA > quantile_N] <- "CNA+"
sum(infer_CNV_T_colMeans$CNA == "CNA+")
#4161 : CNA +
infer_CNV_T_colMeans$CNA[infer_CNV_T_colMeans$CNA <= quantile_N] <- "CNA-"
infer_CNV_T_colMeans[3566, 2] <- "CNA-"
sum(infer_CNV_T_colMeans$CNA == "CNA-")
#9041 : CNA -

infer_CNV_T_colMeans <- cbind("cell" = rownames(infer_CNV_T_colMeans), infer_CNV_T_colMeans)
rownames(infer_CNV_T_colMeans) <- 1:nrow(infer_CNV_T_colMeans)

write.table(infer_CNV_T_colMeans, file="~/Documents/projet_stage/data/inferCNV/infer_CNV_T_colMeans.txt", quote=FALSE, sep="\t", row.names=F)

load("~/Documents/projet_stage/data/inferCNV/infer_cnv_analysis.RData")

#Courbe de densité
plot(density(infer_CNV_N_colMeans$infer_CNV_N_colMeans), main="Average CNV", lwd=2, xlim=c(0, 0.04))
lines(density(infer_CNV_T_colMeans$infer_CNV_T), col="red", lwd=2)
abline(v=quantile_N, lty=2, lwd=2)
legend(0.026, 80, legend=c("Control samples", "Cancer samples", "C99 = 0.0212"), col=c("black", "red", "black"), lty=c(1, 1, 2), cex=0.9, box.lty=0, bg="white")
```

### **Analysis**
#### POOL 
```{r}
#Extract data from seurat object
#head(pool@meta.data, 10)

write.table(pool@meta.data, file="~/Documents/projet_stage/data/analysis_pool_object/metadata_pool_object_orig_ident.tsv", sep="\t", row.names=T, col.names=T, quote=F)

write.table(pool@active.ident, file="~/Documents/projet_stage/data/analysis_pool_object/active_ident_pool_object.tsv", sep="\t", row.names=T, col.names=T, quote=F)
```

#**Addmodulescore**
```{r}
load("~/Documents/projet_stage/data/signatures/geneSignatures.RData")
load("~/Documents/projet_stage/data/data/pool_object.RData")

head(ImmunePop)

#ImmunePop Signatures
pool <- AddModuleScore(pool, ImmunePop, name="module_score_ImmunePop_Sig")

head(pool@meta.data)

plot4 <- FeaturePlot(object=pool, features="module_score_ImmunePop_Sig1", label=TRUE, repel=TRUE)
plot4
plot3 <- VlnPlot(pool, features="module_score_ImmunePop_Sig1", pt.size = FALSE) + theme(legend.position='none')
plot3 <- plot3+labs(title="B lineage")
plot3
plot4+plot3

#Others signatures
pool <- AddModuleScore(pool, otherSig, name="module_score_other_Sig")
head(pool@meta.data)

#Average of add_module_score
metadata <- pool@meta.data
matrix <- aggregate(list(metadata$module_score_other_Sig1, metadata$module_score_other_Sig2, metadata$module_score_other_Sig3, metadata$module_score_other_Sig4, metadata$module_score_ImmunePop_Sig1, metadata$module_score_ImmunePop_Sig2, metadata$module_score_ImmunePop_Sig3, metadata$module_score_ImmunePop_Sig4, metadata$module_score_ImmunePop_Sig5, metadata$module_score_ImmunePop_Sig6, metadata$module_score_ImmunePop_Sig7, metadata$module_score_ImmunePop_Sig8, metadata$module_score_ImmunePop_Sig9, metadata$module_score_ImmunePop_Sig10), by=list(metadata$seurat_clusters), FUN=median)
t(matrix)
names(matrix)[15] <- "avg_module_score_ImmunePop_Sig10"

matrix <- as.data.frame(matrix)
rownames(matrix) <- matrix$seurat_clusters
matrix$seurat_clusters <- NULL
colnames(matrix)
matrix <- matrix[,-9]

names(matrix)<- c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9","10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34")
matrix <- as.data.frame(t(matrix))

write_xlsx(matrix,"~/Documents/projet_stage/data/signatures/matrix_avg_module_score.xlsx")
```

#**Complexheatmap**
```{r}
load("~/Documents/projet_stage/data/signatures/avg_module_score.RData")

library(devtools)
library(usethis)
install_github("jokergoo/ComplexHeatmap")
library(ComplexHeatmap)
library(circlize)

matrix <- as.matrix(matrix)
heatmap <- Heatmap(matrix)
```
