---
title: "Projet_Stage"
author: "Rose-Marie Fraboulet"
output: html_document
---

# **Stage M1 - Analyse de l’hétérogénéité tumorale du cancer du pancréas par des approches single-cell**

```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo=TRUE)
```

## **Loading packages**
```{r loading_packages, include=FALSE}
library(data.table) #large data
library(dplyr) #data manipulation

library(Seurat)
library(patchwork) #plots
library(ggplot2)
library(umap)
library(tsne)

library("writexl")

#scCatch
##install.packages(pkgs = 'devtools')
##devtools::install_github('ZJUFanLab/scCATCH')
library(scCATCH)

#cellasign 
##install.packages("tensorflow")
##install.packages("reticulate")
library(reticulate)
##py_install("tensorflow", pip=T)
library(tensorflow)
##install_tensorflow(extra_packages = "tensorflow-probability")
#devtools::install_github("Irrationone/cellassign")
library(SingleCellExperiment)
library(cellassign)
```

## **Loading dataset**
### Pool
```{r, echo=FALSE}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/code/projet.RData")
#load("/home/rose/Documents/projet.RData")
```

### Pool + metadata (labels) / Pool2
```{r, echo=FALSE}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre #2/Stage M1/Projet/data/peng_data")

#metadata <- fread("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/peng_data/all_celltype.txt", data.table=FALSE)

#Rows are cell names and the columns are additional metadata fields.
##rownames(metadata) <- metadata$cell.name
##head(metadata)
##metadata$cell.name <- NULL

#pool2 <- CreateSeuratObject(counts = counts, meta.data = metadata)
```

### Pool.init 
```{r}
#load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/init.RData")
#load("/home/rose/Documents/init.RData")
```

### All objects
```{r}
load("/home/rose/Documents/data_clusters.RData")
load("~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/code/data_clusters.RData")
pool
pool2
pool.init
```

## **Standard pre-processing workflow - Data filtering**
### Calculate the percentage of mitochondrial genes 
```{r}
pool[["percent.mt"]] <- PercentageFeatureSet(object = pool, pattern = "^MT-")
#[[""]] : add a column "percent.mt" in the pool object 
#number of mitochondrial genes for one cell / total number of genes in that cell * 100
```

```{r}
head(x = pool@meta.data, 10)
#The total number of molecules detected within a cell : nCount (nUMIs)
#The number of genes detected in each cell : nFeature (nGenes)
#The percentage of reads that map to the mitochondrial genome : percent.mt 
```

### Graphical visualization 
#### Violin plot
```{r}
VlnPlot(pool, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

#### Plot
```{r}
plot1 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot1

plot2 <- FeatureScatter(pool, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2
```

#### Low quality cells
<200 genes/cell and >10% mitochondrial genes were excluded 
```{r}
pool <- subset(pool, subset = nFeature_RNA > 200 & percent.mt < 10)
pool
```

## **Normalizing data - logNormalize**

## Normalizing the data
```{r}
#Normalizes the gene expression for each cell by the total expression, multiplies this by a scale factor (10 000) and log-transforms the result.
pool <- NormalizeData(pool, normalization.method = "LogNormalize", scale.factor=10000)
```

## Identification of highly variable features
```{r}
#Calculates highly variable genes - Highly expressed in some cells, and lowly expressed in others
pool <- FindVariableFeatures(object=pool, selection.method = "vst", nfeatures = 2000)
head(x = HVFInfo(object = pool))

#The 10 most higly variable genes
top10.variablegenes <- head(VariableFeatures(pool), 10)
top10.variablegenes

#plot3 <- VariableFeaturePlot(pool)
#plot3

plot4 <- LabelPoints(plot = plot3, points = top10_variablegenes, repel = TRUE, xnudge=0, ynudge=0)
plot4
```

## **Scaling data**
```{r}
# Pre-processing step before PCA
pool <- ScaleData(pool)
```

## **Perform linear dimensional reduction**
### Dataset
```{r}
pool <- RunPCA(pool, features = VariableFeatures(object = pool))
#print(pool[["pca"]], dims = 1:5, nfeatures = 5)
```

### Visualization
#### VizDimReduction
```{r}
VizDimLoadings(pool, dims = 1:2, reduction = "pca")
```

#### DimPlot
```{r}
DimPlot(pool, reduction = "pca", label=TRUE, repel=TRUE)
```

#### DimHeatMap
```{r}
DimHeatmap(pool, dims = 1:3, cells = 500, balanced = TRUE)
#black = 0, purple = low, yellow = high expression
```

## **Determine the ‘dimensionality’ of the dataset**
```{r}
ElbowPlot(object=pool, ndims = 50, reduction = "pca")
```

## **Cluster the cell**
### Shared Nearest Neighbor (SNN) Graph
```{r}
#Determine the k-nearest neighbors of each cell
#PCs 1 to 10 were used for graph-based clustering

pool <- FindNeighbors(pool, reduction='pca', dims = 1:10)

pool <- FindClusters(pool, resolution = 1) 
#Res = 0.8 for PDAC samples 
#Res = 1 for pooled cells from PDACs and controls
```

#### Visualization 
```{r}
head(Idents(pool), 10)

#Which cluster has been determined for each cell
head(pool@meta.data, 10)

#Delete a column :  pool[['RNA_snn_res.1.5']] <- NULL
```

## **Run non-linear dimensional reduction**

### Uniform Manifold Approximation and Projection (UMAP) 
```{r, include=FALSE}
pool <- RunUMAP(pool, dims = 1:10)
```

### T-distributed stochastic neighbor embedding (tSNE)
```{r, results="hide"}
# pool <- RunTSNE(pool, dims.use = 1:10)
# 
# plot8 <- DimPlot(object = pool, reduction = "tsne", label=TRUE, repel=TRUE, group.by="orig.ident")
# plot9 <- DimPlot(object = pool, reduction = "tsne", label=TRUE, repel=TRUE, group.by="cluster") + NoLegend()
# plot8 + plot9
```

## **Finding differentially expressed features (cluster biomarkers)**

### **Using markers cited in the article**
#### Find markers for cluster 1
```{r}
cluster1_markers <- FindMarkers(pool, ident.1 = 1)
head(cluster1_markers, n = 10)
#pct.1 = quel pourcentage des cellules dans la condition 1 montrent une certaine expression pour ce gène
#pct.2 = quel pourcentage des cellules dans la condition 2 montrent une certaine expression pour ce gène
#avg_logFC = combien l'expression de ce gène est plus ou moins élevée dans le cluster choisi entre les conditions
```

#### Find markers for every cluster compared to all remaining cells
```{r}
pool_markers <- FindAllMarkers(pool, only.pos = TRUE)
pool_markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
head(pool_markers)
```

#### Visualization 
```{r}
#For a specific marker
feature <- c("MET")
```

##### Violin plot
```{r}
VlnPlot(pool, features = feature, pt.size= FALSE, group.by="cluster")
```

##### Feature plot
```{r}
FeaturePlot(pool, features = feature, label=TRUE, repel=TRUE)
#FeaturePlot(pool, features = feature, reduction = "tsne", label=TRUE, repel=TRUE)
```

##### DoHeatMap
```{r}
DoHeatmap(subset(pool, downsample = 10), features=feature, size=3)
```

#### Top 100 differentially expressed genes for each cluster
```{r, results="hide"}
top100.markers <- pool.markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_logFC)
head(top100.markers)
#DoHeatmap(pool, features = top100_markers$gene, size=3)
```

```{r}
# .xlsx file of the top 100 markres for each cluster

#write_xlsx(top100.markers, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/Data/top100_markers.xlsx")
```

#### **Assigning cell type identity to clusters**
```{r}
experiment.merged <- RenameIdents(object = pool, '30'= '0', '31'= '0', '26'= '0', '6' = '1', '8' = '1', '9'= '2', '20'= '2', '25'= '2', '24'= '2', '25'= '2', '18'= '2', '14'= '2', '13'= '3', '29'= '3', '27'= '4', '34'= '4', '12'= '4', '5'= '4', '33'= '4', '32'= '4', '17'='22', '11'='10', '15'='10', '21'='10', '16'='7', '28'='7', '19'='7', '33'='7')

#DimPlot(object = experiment.merged, pt.size=0.5, label = T, reduction = "umap")
#DimPlot(object = experiment.merged, pt.size=0.5, label = T, reduction = "tsne")

new.cluster.ids <- c("Stellate", "Ductall cells 1", "Ductall cells 2", "Macrophages", "Endothelial cells", "Acinar cells", "Fibroblasts", "T/B cells", "Endocrine cells")
names(new.cluster.ids) <- levels(experiment.merged)
pool <- RenameIdents(experiment.merged, new.cluster.ids)

DimPlot(pool, reduction = "umap", pt.size = 0.5, repel=TRUE, label= TRUE) + NoLegend()
#DimPlot(pool, reduction = "tsne", pt.size = 0.5, repel=TRUE, label= TRUE) + NoLegend()
```

### **Clustering tools - pool.init**
```{r}
#Using pool.init = pool data without the new clusters id found by known markers
```

#### **BuildClusterTree**
```{r}
#cluster_tree = BuildClusterTree(pool.init)
#Tool(object = cluster_tree, slot = 'BuildClusterTree')
#PlotClusterTree(object = cluster_tree)
```

#### **scCATCH**
```{r}
clu_markers_sccatch <- findmarkergenes(pool.init, species = 'Human', cluster = 'All', match_CellMatch = TRUE, cancer = NULL, tissue = 'Pancreas', cell_min_pct = 0.25, logfc = 0.25, pvalue = 0.05)
#large datasets > 10,000 cells or > 15 clusters

clu_ann_sccatch <- scCATCH(clu_markers_sccatch$clu_markers, species = 'Human', cancer = NULL, tissue = 'Pancreas')

#write_xlsx(clu_ann_sccatch, "~/Desktop/stage_rose/clu_ann_sccatch.xlsx")
```

#### **cellasign**
```{r}

```

### **Using cell labels - pool2**
```{r}
plot5 <- DimPlot(pool2, reduction = "umap", label=TRUE, repel=TRUE, group.by="orig.ident")
plot5

experiment.merged2 <- RenameIdents(object = pool2, 'N11'='N1', 'N10'='N1', 'N9'='N1', 'N8'='N1', 'N7'='N1', 'N6'='N1', 'N5'='N1', 'N4'='N1', 'N3'='N1', 'N2'='N1', 'T2'='T1', 'T3'='T1','T4'='T1','T5'='T1','T6'='T1','T7'='T1','T8'='T1','T9'='T1','T10'='T1','T11'='T1','T12'='T1','T13'='T1','T14'='T1','T15'='T1','T16'='T1','T17'='T1','T18'='T1','T19'='T1','T20'='T1','T21'='T1','T22'='T1','T23'='T1','T24'='T1')
#plot6 <-DimPlot(object = experiment.merged2, pt.size=0.5, label = T, reduction = "umap")
new.orig_ident.ids <- c("Control samples", "Cancer samples")
names(new.orig_ident.ids) <- levels(experiment.merged2)
pool2 <- RenameIdents(experiment.merged2, new.orig_ident.ids)

plot6 <- DimPlot(pool2, reduction = "umap", pt.size = 0.5, cols=c("#99CCFFE6", "coral"))
plot6 

plot7 <- DimPlot(pool2, reduction = "umap", label=TRUE, repel=TRUE, group.by="cluster") + NoLegend()
plot7
```

### **Calculating the average gene expression within a cluster**
```{r}
Idents(pool2) <- "cluster"
#Idents(pool2) <- "orig.ident"

#How many cells are in each cluster
table(Idents(pool2))

#How many cells are in each sample
table(pool2$orig.ident)

#How many cells per sample
table(Idents(pool2), pool2$orig.ident)

cluster.averages <- AverageExpression(pool2)
head(cluster.averages[["RNA"]][, 1:5])

#Add cluster.averages to the pool2 object
orig.levels <- levels(pool2)
Idents(pool2) <- gsub(pattern = " ", replacement = "_", x = Idents(pool2))
orig.levels <- gsub(pattern = " ", replacement = "_", x = orig.levels)
levels(pool2) <- orig.levels
cluster.averages <- AverageExpression(pool2, return.seurat = TRUE)
cluster.averages

CellScatter(cluster.averages, cell1 = "Ductal_cell_type_1", cell2 = "Ductal_cell_type_2")

DoHeatmap(cluster.averages, features = unlist(TopFeatures(pool2[["pca"]], balanced = TRUE)), size = 3, 
    draw.lines = FALSE)
```


```{r}
##Contengency table
orig.ident <- pool2$orig.ident
clusters <- pool.init$seurat_clusters

contengency.table <- data.frame(orig.ident, clusters)

table <- data.frame(metadata, orig.ident, clusters)

#write_xlsx(contengency.table, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/supplementary_information/contengency_table.xlsx")

#write_xlsx(table, "~/Université/BIOLOGIE/Master Bioinformatique/M1 Bioinformatique/Semestre 2/Stage M1/Projet/data/supplementary_information/table.xlsx")
```


